### PR TITLE
feat(scheduler): flexible time windows

### PR DESCRIPTION
**Pull request by sakurai-ryo**
This PR adds support for configuring flexible time windows.

## Description
Currently, users cannot configure the `flexibleTimeWindow` feature in the Scheduler construct.
This feature enhances flexibility and reliability, allowing tasks to be invoked within a defined time window. 
https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-schedule-flexible-time-windows.html

CloudFormation allows users to take advantage of this feature as follows.
With this template,  it will invokes the target within 10 minutes after the scheduled time.
```yaml
Resources:
  Schedule:
    Type: AWS::Scheduler::Schedule
    Properties: 
      FlexibleTimeWindow: 
        Mode: "FLEXIBLE" # or "OFF"
        MaximumWindowInMinutes: 10 # between 1 and 1440
      Name: "sample-schedule"
      ScheduleExpression: "cron(0 9 * * ? *)"
      State: "ENABLED"
      Target:
        Arn: hoge
        RoleArn: hoge
```

### add Enum indicating flexible time window mode
Currently there are only two modes, FLEXIBLE and OFF, so there is no problem using boolean instead of enum.
But I think it's better to use Enum to prepare for future expansion.
https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-flexibletimewindow.html

### add property to `ScheduleProps` interface
`flexibleTimeWindowMode` property defaults to `OFF` to avoid a breaking change.
```ts
interface ScheduleProps {
  // ....
  /**
   * Determines whether the schedule is invoked within a flexible time window.
   *
   * @see https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-schedule-flexible-time-windows.html
   *
   * @default - FlexibleTimeWindowMode.OFF
   */
  readonly flexibleTimeWindowMode?: FlexibleTimeWindowMode;

  /**
   * The maximum time window during which the schedule can be invoked.
   *
   * @default - Required if flexibleTimeWindowMode is FLEXIBLE.
   */
  readonly maximumWindowInMinutes?: Duration;
}
```

### set the added property to `CfnSchedule` construct
Basically, just set the values as documented, but with the following validations.
- If `flexibleTimeWindowMode` is `FLEXIBLE`
  - `maximumWindowInMinutes` must be specified
  - `maximumWindowInMinutes` must be set from 1 to 1440 minutes

https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-flexibletimewindow.html

In addition, I added some unit tests and integ-tests.

### others
- fixed typo in README
  -  `customizeable` => `customizable`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by go-to-k — CHANGES_REQUESTED** [review]
Thanks for this PR!

I would like to add a new interface to `FlexibleTimeWindow`, what do you think?

Even if there are more properties associated with `FlexibleTimeWindow` in the future, it might not be too complicated. It might also be simpler than validation by annotation (as it is now).

**Comment by sakurai-ryo** [comment]
@go-to-k 
Thank you for your feedback!
Initially, I had implemented this by defining a new interface, but after reviewing the section outlined in the `DESIGN_GUIDELINES`, I decided to go with the current implementation (although the naming of the properties is somewhat confusing).
https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#flat

I'm wondering if the addition of the new interface is in line with this guideline.

**Comment by go-to-k** [comment]
Thanks, I see. Certainly correct. Let's go with this. 
Please wait a moment so I can look at the other parts.

File: packages/@aws-cdk/aws-scheduler-alpha/test/schedule.test.ts [hunk 1]
```diff
@@ -128,4 +128,71 @@ describe('Schedule', () => {
...
+      expect(() => {
+        new Schedule(stack, 'TestSchedule', {
+          schedule: expr,
+          target: new SomeLambdaTarget(func, role),
+          flexibleTimeWindowMode: FlexibleTimeWindowMode.FLEXIBLE,
+        });
+      }).toThrow('maximumWindowInMinutes must be provided when flexibleTimeWindowMode is set to FLEXIBLE');
+    });
+
+    test('throw error when maximumWindowInMinutes is more than 1440', () => {
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-alpha/test/schedule.test.ts:176]: It is not big deal but...

```suggestion
    test('throw error when maximumWindowInMinutes is greater than 1440', () => {
```

**Review by go-to-k — CHANGES_REQUESTED** [review]
I commented some points. Please feel free to ask me if you need anything.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 2]
```diff
@@ -306,4 +340,25 @@ export class Schedule extends Resource implements ISchedule {
...
+      throw new Error('maximumWindowInMinutes must be provided when flexibleTimeWindowMode is set to FLEXIBLE');
+    }
+    if (maximumWindowInMinutes.toMinutes() < 1 || maximumWindowInMinutes.toMinutes() > 1440) {
+      throw new Error(`maximumWindowInMinutes must be between 1 and 1440, got ${maximumWindowInMinutes.toMinutes()}`);
+    }
+    return {
+      mode: flexibleTimeWindowMode,
+      maximumWindowInMinutes: maximumWindowInMinutes.toMinutes(),
+    };
+  }
```
Comment (go-to-k) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:363]: Thanks for the fix!

Sorry, this one might be better.

```suggestion
  private renderFlexibleTimeWindow(
    flexibleTimeWindowMode?: FlexibleTimeWindowMode, maximumWindowInMinutes?: Duration,
  ): CfnSchedule.FlexibleTimeWindowProperty {
    const mode = flexibleTimeWindowMode ?? FlexibleTimeWindowMode.OFF;

    if (mode === FlexibleTimeWindowMode.OFF) {
      return {
        mode,
      };
    }

    if (!maximumWindowInMinutes) {
      throw new Error('maximumWindowInMinutes must be provided when flexibleTimeWindowMode is set to FLEXIBLE');
    }
    if (maximumWindowInMinutes.toMinutes() < 1 || maximumWindowInMinutes.toMinutes() > 1440) {
      throw new Error(`maximumWindowInMinutes must be between 1 and 1440, got ${maximumWindowInMinutes.toMinutes()}`);
    }
    return {
      mode,
      maximumWindowInMinutes: maximumWindowInMinutes.toMinutes(),
    };
  }
```

**Review by go-to-k — CHANGES_REQUESTED** [review]

**Review by go-to-k — APPROVED** [review]
Thanks, looks good.

**Comment by sakurai-ryo** [comment]
@go-to-k 
Thanks for the quick review!

**Review by scanlonp — CHANGES_REQUESTED** [review]
Hey @sakurai-ryo, apologies we are getting around to this a little late. I think this case is a great opportunity for an Enum like class! For example
```
new scheduler.Schedule(stack, 'UseFlexibleTimeWindow', {
  schedule: expression,
  target: target,
  flexibleTimeWindow: scheduler.FlexibleTimeWindowMode.Flexible(cdk.Duration.minutes(10)),
});
```
and 
```
new scheduler.Schedule(stack, 'UseFlexibleTimeWindow', {
  schedule: expression,
  target: target,
  flexibleTimeWindow: scheduler.FlexibleTimeWindowMode.Off(),
});
```
This will tightly couple the mode to the properties it needs! I think this will give the best user experience and be extensible.

You can certainly improve my naming as well!

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 3]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
...
 }
 
+/**
+ * A time window during which EventBridge Scheduler invokes the schedule.
+ */
+export class FlexibleTimeWindowMode {
+  /**
+   * FlexibleTimeWindow is disabled.
+   */
+  public static off(): FlexibleTimeWindowMode {
```
Comment (sakurai-ryo) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:69]: I changed it to camelCase because it resulted in an error during cdk-build.
```
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "@aws-cdk/aws-scheduler-alpha.FlexibleTimeWindowMode.FLEXIBLE" to "flexible"
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "@aws-cdk/aws-scheduler-alpha.FlexibleTimeWindowMode.OFF" to "off"
```

**Comment by sakurai-ryo** [comment]
@scanlonp 
No problem at all! Thank you for the review!
I've made the changes. Please review them again.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 4]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
...
+   * Determines whether the schedule is invoked within a flexible time window.
+   */
+  public readonly mode: string;
+
+  /**
+   * The maximum time window during which the schedule can be invoked.
+   *
+   * Must be between 1 to 1440 minutes.
+   *
+   * @default - Required if mode is FLEXIBLE.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:93]: This isn't the correct use of the `@default` tag. Should describe the default if mode is not flexible

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 5]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
...
+    if (maximumWindowInMinutes.toMinutes() < 1 || maximumWindowInMinutes.toMinutes() > 1440) {
+      throw new Error(`maximumWindowInMinutes must be between 1 and 1440, got ${maximumWindowInMinutes.toMinutes()}`);
+    }
+    return new FlexibleTimeWindowMode('FLEXIBLE', maximumWindowInMinutes);
+  }
+
+  /**
+   * Determines whether the schedule is invoked within a flexible time window.
+   */
+  public readonly mode: string;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:86]:
```suggestion
  public readonly mode: 'OFF' | 'FLEXIBLE';
```

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 6]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
...
+   * FlexibleTimeWindow is disabled.
+   */
+  public static off(): FlexibleTimeWindowMode {
+    return new FlexibleTimeWindowMode('OFF');
+  }
+
+  /**
+   * FlexibleTimeWindow is enabled.
+   */
+  public static flexible(maximumWindowInMinutes: Duration): FlexibleTimeWindowMode {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:76]:
```suggestion
  public static flexible(maxWindow: Duration): FlexibleTimeWindowMode {
```

It doesn't need to be in minutes anymore cuz we're supplying a Duration

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 7]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
   readonly retryAttempts?: number;
 }
 
+/**
+ * A time window during which EventBridge Scheduler invokes the schedule.
+ */
+export class FlexibleTimeWindowMode {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:65]:
```suggestion
export class TimeWindow {
```

Comment (kaizencc) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:65]: Will be nicer to write `TimeWindow.flexible()`

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 8]
```diff
@@ -59,6 +59,47 @@ export interface ScheduleTargetProps {
...
+  public readonly mode: string;
+
+  /**
+   * The maximum time window during which the schedule can be invoked.
+   *
+   * Must be between 1 to 1440 minutes.
+   *
+   * @default - Required if mode is FLEXIBLE.
+   */
+  public readonly maximumWindowInMinutes?: Duration;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:95]:
```suggestion
  public readonly maxWindow?: Duration;
```

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 9]
```diff
@@ -114,6 +156,15 @@ export interface ScheduleProps {
...
   readonly key?: kms.IKey;
+
+  /**
+   * A time window during which EventBridge Scheduler invokes the schedule.
+   *
+   * @see https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-schedule-flexible-time-windows.html
+   *
+   * @default FlexibleTimeWindowMode.off()
+   */
+  readonly flexibleTimeWindow?: FlexibleTimeWindowMode;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:167]:
```suggestion
  readonly timeWindow?: TimeWindow;
```

**Review by kaizencc — CHANGES_REQUESTED** [review]
Hi @sakurai-ryo, I have some naming convention nits here. But otherwise looks good

**Comment by sakurai-ryo** [comment]
@kaizencc 
Thanks! I fixed them.

**Review by kaizencc — CHANGES_REQUESTED** [review]
Oh shoot! Merge conflicts with your other PR that got merged :). Do you mind fixing those conflicts @sakurai-ryo? Everything else looks good.

**Comment by sakurai-ryo** [comment]
@kaizencc 
fixed those conflicts 👍

**Review by kaizencc — APPROVED** [review]
